File: src/drivers/term.cpp
#include <drivers/term.hpp>
#include <unifont.h>
#include <limine.h>
#include <limine_terminal/term.h>
#include <lib/modules.hpp>
#include <io/stdio.hpp>
#include <lib/logger.hpp>
#include <drivers/fs/devtmpfs.hpp>
using namespace term;

__attribute__((section(".requests")))
static limine_framebuffer_request fb_request = { LIMINE_FRAMEBUFFER_REQUEST, 2, nullptr };

struct fb_cdev_t : devtmpfs::cdev_t {
    fb_cdev_t(void *address, std::size_t size) : fb{static_cast<char*>(address)}, size{size} {}
    ~fb_cdev_t() = default;
    
    ssize_t read(void *buffer, size_t size, off_t off) {
        size = std::min(this->size - off, size);
        memcpy(fb + off, buffer, size);
        return size;
    }

    ssize_t write(void *buffer, size_t size, off_t off) {
        size = std::min(this->size - off, size);
        memcpy(buffer, fb + off, size);
        return size;
    }

private:
    char *fb;
    std::size_t size;
};

struct tty_cdev_t : devtmpfs::cdev_t {};

std::vector<term_t*> terms{fb_request.response->framebuffer_count};

class term_writer {
public:
    void append(char c) {
        for (const auto& t : terms)
            term_write(t, &c, 1);
    }

    void append(std::string_view s) {
        for (const auto& t : terms)
            term_write(t, s.data(), s.size());
    }

    void append(const char *s) { append(std::string_view{s}); }
};

static term_writer *writer;

void term::init() {
    auto font_file = mod::find("font");

    font_t font = {
        reinterpret_cast<uintptr_t>(font_file->address),
        UNIFONT_WIDTH,
        UNIFONT_HEIGHT,
        TERM_FONT_SPACING,
        TERM_FONT_SCALE_X,
        TERM_FONT_SCALE_Y
    };

    style_t style = {
        TERM_ANSI_COLOURS,
        TERM_ANSI_BRIGHT_COLOURS,
        TERM_BACKGROUND,
        TERM_FOREGROUND_BRIGHT,
        TERM_BACKGROUND_BRIGHT,
        TERM_FOREGROUND_BRIGHT,
        TERM_MARGIN,
        TERM_MARGIN_GRADIENT
    };

    background_t back = {
        nullptr,
        IMAGE_TILED,
        TERM_BACKDROP
    };

    for (std::size_t i = 0; i < fb_request.response->framebuffer_count; i++) {
        auto framebuffer = fb_request.response->framebuffers[i];

        terms[i] = term_init(
            {
                reinterpret_cast<uintptr_t>(framebuffer->address),
                framebuffer->width,
                framebuffer->height,
                framebuffer->pitch,
                framebuffer->red_mask_size,
                framebuffer->red_mask_shift,
                framebuffer->green_mask_size,
                framebuffer->green_mask_shift,
                framebuffer->blue_mask_size,
                framebuffer->blue_mask_shift
            },
            font,
            style,
            back
        );
        terms[i]->cursor_enabled = false;
    }
    clear();

    writer = new term_writer;
    io::print = [](const char *f, frg::va_struct *args) {
        return frg::printf_format(io::char_printer{*writer, args}, f, args);
    };

    logger::info("TERM", "Initialized terminal");
}

void term::register_devs() {
    for (std::size_t i = 0; i < fb_request.response->framebuffer_count; i++) {
        auto fb = fb_request.response->framebuffers[i];
        devtmpfs::add_dev(std::string{"fb"} + std::to_string(i), new fb_cdev_t{fb->address, fb->pitch * fb->height}, S_IFCHR | S_IRWXU);
    }
}

void term::clear() {
    for (auto& t : terms)
        t->clear(t, true);
}

File: src/drivers/term.zig
const limine = @import("limine");
const flanterm = @import("flanterm");
const kernel = @import("kernel");
const mem = kernel.lib.mem;
const log = @import("std").log.scoped(.term);
const ArrayList = @import("std").ArrayList;
const std = @import("std");
export var FB_REQUEST: limine.FramebufferRequest = .{};

var terms: []*flanterm.struct_flanterm_context = &.{};

pub var Task = kernel.Task{
    .name = "Framebuffer",
    .init = init,
    .dependencies = &.{
        .{ .task = &mem.KHeapTask },
    },
};

fn init() bool {
    const framebuffers = (FB_REQUEST.response orelse return false);
    var term_arr = ArrayList(*flanterm.struct_flanterm_context).init(mem.kheap.allocator());
    for (0..framebuffers.framebuffer_count) |i| {
        const fb = framebuffers.framebuffers_ptr[i];
        const ctx = flanterm.flanterm_fb_init(
            malloc,
            free,
            @ptrCast(@alignCast(fb.address)),
            fb.width,
            fb.height,
            fb.pitch,
            fb.red_mask_size,
            fb.red_mask_shift,
            fb.green_mask_size,
            fb.green_mask_shift,
            fb.blue_mask_size,
            fb.blue_mask_shift,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            null,
            0,
            0,
            1,
            0,
            0,
            0,
        );
        term_arr.append(ctx) catch return false;
    }
    terms = term_arr.toOwnedSlice() catch return false;
    return true;
}

pub fn write(bytes: []const u8) !usize {
    if (terms.len == 0) return 0;
    for (terms) |t| flanterm.flanterm_write(t, bytes.ptr, bytes.len);
    return bytes.len;
}

fn malloc(size: usize) callconv(.C) ?*anyopaque {
    return @ptrCast(mem.kheap.allocator().alloc(u8, size) catch @panic("OOM"));
}

fn free(ptr: ?*anyopaque, size: usize) callconv(.C) void {
    const arr: [*]u8 = @ptrCast(ptr orelse @panic("Free called with NULL"));
    mem.kheap.allocator().free(arr[0..size]);
}
